### PR TITLE
Xliff merge tool

### PR DESCRIPTION
**Pull request by admd**
## What does this PR change?

A simple tool to manage xliff files

## GUI diff

No difference.

Before:

After:

- [ ] **DONE**

## Documentation
- No documentation needed: **add explanation. This can't be used if there is a GUI diff**
- Documentation issue was created: [Link for SUSE Manager contributors](https://github.com/SUSE/spacewalk/issues/new?template=ISSUE_TEMPLATE_DOCUMENTATION.md&labels=documentation&projects=SUSE/spacewalk/31), [Link for community contributors](https://github.com/uyuni-project/uyuni-docs/issues/new).
- (OPTIONAL) [Documentation PR](https://github.com/uyuni-project/uyuni-docs/pulls)

- [ ] **DONE**

## Test coverage
- No tests: **add explanation**
- Unit tests were added
- Cucumber tests were added

- [ ] **DONE**

## Links

Fixes #
Tracks # **add downstream PR, if any**

- [ ] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: scripts/translation/xliffmerger.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+def delete_trans_units(body_element, trans_units, units_to_remove):
+    to_delete = [tr_unit for tr_unit in trans_units if tr_unit.attrib['id'] in units_to_remove]
+    for item in to_delete:
+        body_element.remove(item)
+
+
+def add_trans_units(body_element, trans_units, units_to_add):
+    to_add = [tr_unit for tr_unit in trans_units if tr_unit.attrib['id'] in units_to_add]
+    for item in to_add:
+        item.set('state', 'new')
```
Comment (mcalmer) [inline scripts/translation/xliffmerger.py:18]: `state` is an attribute for "&lt;target&gt;". Here it seems you set it on &lt;trans-unit&gt; .
But I think we do not need to set state=new as a missing &lt;target&gt; has the same effect

File: scripts/translation/xliffmerger.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        body_element.append(item)
+
+
+def update_trans_units(body_element, trans_units, id_new_value_dict):
+    to_update = [tr_unit for tr_unit in trans_units if tr_unit.attrib['id'] in id_new_value_dict]
+    for trans_unit in to_update:
+        trans_unit.set('state', 'needs-adaptation')
```
Comment (mcalmer) [inline scripts/translation/xliffmerger.py:25]: `state` is an attribute for "&lt;target&gt;". Here it seems you set it on &lt;trans-unit&gt;

http://docs.oasis-open.org/xliff/v1.2/os/xliff-core.html#state

File: scripts/translation/xliffmerger.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        trans_unit[0].text = id_new_value_dict[trans_unit.attrib['id']]
```
Comment (mcalmer) [inline scripts/translation/xliffmerger.py:26]: Can we be sure that trans_unit[0] is element "source"? I think we better pick the source element explicit.

File: scripts/translation/xliffmerger.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    # add the 'needs-adaptation' state attribute
+    logging.info("########## UPDATE THE CHANGED TRANS_UNITS ##########")
+    # Get again so we get the updated list after deletion/addition
+    translation_trans_units = list(translation_body_element)
+
+    logging.debug(f'{len(original_trans_units)} trans_units in original file {original_file}')
+    logging.debug(f'{len(translation_trans_units)} trans_units in original file {original_file}')
+    if len(translation_trans_units) == len(original_trans_units):
+        trans_units_sources = {tr_unit.attrib['id']: tr_unit[0].text for tr_unit in translation_trans_units}
+        org_trans_units_sources = {tr_unit.attrib['id']: tr_unit[0].text for tr_unit in original_trans_units}
```
Comment (mcalmer) [inline scripts/translation/xliffmerger.py:66]: Can we be sure that tr_unit[0] is element "source"? I think we better pick the source element explicit.

File: scripts/translation/xliffmerger.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        to_update = {k: org_trans_units_sources[k] for k,_ in set(org_trans_units_sources.items()) - set(trans_units_sources.items())}
+        logging.info(f' These (id,new_source_value) -> {to_update}, will be updated in the {translation_file} ')
+        update_trans_units(translation_body_element,translation_trans_units,to_update)
+
```
Comment (mcalmer) [inline scripts/translation/xliffmerger.py:70]: What about `else:` (if len(translation_trans_units) == len(original_trans_units):)

I think we should at least write a message

File: scripts/translation/xliffmerger.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    # file_tag = root_node[0]
+    body_element = root_node[0][0]
+    return body_element, list(body_element)
+
+
+ET.register_namespace('', "urn:oasis:names:tc:xliff:document:1.1")
+ET.register_namespace('xyz', "urn:appInfo:Items")
+ET.register_namespace('xsi', "http://www.w3.org/2001/XMLSchema-instance")
+
+files = os.listdir('.')
```
Comment (mcalmer) [inline scripts/translation/xliffmerger.py:87]: Would be good if we could provide the working directory as command line parameter.

**Comment by admd** [comment]
@mcalmer 
`xliffmerge` tool is ready to test. I have added another small script `trans-units-extractor` script to extract `trans-units` from `<group>` element and then remove that parent `<group>` element. It works but doesn't keep the order which makes diff pretty bad. I will try to see if we can somehow keep the order.

**Comment by mcalmer** [comment]
@parlt91 will take over now.

**Comment by admd** [comment]
The order has been fixed. Now the only thing which is kinda bad is that during modifying, our tools remove the comments if there are any in that translation file. I will be looking into hacking around this part and see if somehow we can keep the comments while using the standard python library. There are some third parties libraries that take care of comments as well but I would prefer to stick with the standard library for the time being.

**Comment by admd** [comment]
Pascal if you get some free time, it would be cool if you could play with these 2 small scripts. Let me know if you need any information. @parlt91

**Comment by parlt91** [comment]
@admd Sure. I should have some time to check them out tomorrow.

**Comment by parlt91** [comment]
@admd So, I played around with the scripts a bit.
I could not find any issues with adding, removing or updating the strings.

I really like adding the `new` and `needs-adaptation` flags :+1: 
I wonder if we could add the`new` flag to all strings that still lack translation? (they should be missing the `<target>` in this case)

**Comment by admd** [comment]
Thank you @parlt91 for having a look. 

> I wonder if we could add the`new` flag to all strings that still lack translation? (they should be missing the `<target>` in this case)
It was based on the assumption that weblate takes care of it, if <target> element is missing, weblate automatically assumes that this needs translation. Ofcourse, I could be wrong as I haven't tested and in that case, we will need to add the element as you suggested. 

BTW I have added a commit and now hopefully comments will be kept and not removed.

File: scripts/translation/xliffmerger.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/python
+import xml.etree.ElementTree as ET
+import os
+import logging
+import argparse
+argparser = argparse.ArgumentParser()
+argparser.add_argument("path", help="Path to look into for translation files!")
+args = parser.parse_args()
```
Comment (parlt91) [inline scripts/translation/xliffmerger.py:8]: `parser` -> `argparser`

File: scripts/translation/xliffmerger.py [hunk 8]
```diff
@@ -40,8 +47,9 @@ def update_trans_units(body_element, trans_units, id_new_value_dict):
 
 
 def process(original_file, translation_file):
-    xml_tree_original = ET.parse(original_file)
-    xml_tree_translation = ET.parse(translation_file)
+    parser = ET.XMLParser(target=CommentedTreeBuilder())
```
Comment (parlt91) [inline scripts/translation/xliffmerger.py:50]: Using the same instance of `parser` two parse both files does not seem to work for me

```
INFO:root:
processing StringResource_ca.xml
Traceback (most recent call last):
  File "/home/pascal/git/uyuni/scripts/translation/xliffmerger.py", line 123, in <module>
    process(original, translation)
  File "/home/pascal/git/uyuni/scripts/translation/xliffmerger.py", line 53, in process
    xml_tree_original = ET.parse(original_file, parser=parser)
  File "/usr/lib64/python3.6/xml/etree/ElementTree.py", line 1196, in parse
    tree.parse(source, parser)
  File "/usr/lib64/python3.6/xml/etree/ElementTree.py", line 603, in parse
    parser.feed(data)
xml.etree.ElementTree.ParseError: parsing finished: line 23921, column 8
```

Comment (parlt91) [reply scripts/translation/xliffmerger.py:50]: Creating two parsers seems to work
```
parser1 = ET.XMLParser(target=CommentedTreeBuilder())
parser2 = ET.XMLParser(target=CommentedTreeBuilder())
xml_tree_original = ET.parse(original_file, parser=parser1)
xml_tree_translation = ET.parse(translation_file, parser=parser2)
```
I have no idea why though

File: scripts/translation/xliffmerger.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        org_trans_units_srcs = {tr_unit.attrib['id']: tr_unit.find('d:source', ns).text for tr_unit in orig_trans_units}
+        to_update = {k: org_trans_units_srcs[k] for k, _ in
+                     set(org_trans_units_srcs.items()) - set(trans_units_srcs.items())}
+        logging.info(f' These (id,new_source_value) -> {to_update}, will be updated in the {translation_file} ')
+        update_trans_units(translation_body_element, trans_trans_units, to_update)
+    else:
+        logging.info("Something went wrong, this should not have happend!")
+
+
+    #xml_tree_translation.write(translation, encoding='utf-8', xml_declaration=True)
```
Comment (parlt91) [inline scripts/translation/xliffmerger.py:96]: Should be un-commented before merging

File: scripts/translation/xliffmerger.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+
+def get_trans_units(xml_tree):
+    root_node = xml_tree.getroot()
+    # file_tag = root_node[0]
+    file_tag = root_node.find('d:file', ns)
+    # body_element = root_node[0][0]
```
Comment (parlt91) [inline scripts/translation/xliffmerger.py:103]: Also make sure to remove unneeded comments

**Review by parlt91 — CHANGES_REQUESTED** [review]
Overall looks already pretty good.
Preserving the comments seems to work fine.
The known issues of messing with the indentation and the order of the translations still persist, but they don't break functionality, so I could personally life with that.

**Comment by admd** [comment]
Hey Pascal, thank you for the review. There were quite a few facepalms :), this is what happens with last-minute changes. I have made the suggested changes. @parlt91

**Review by parlt91 — APPROVED** [review]
LGTM now. Thank you :+1:

**Pull request merged**